File: packages/go/domains.go
package axene

import (
	"context"
	"net/url"
)

// Domains is the domains resource, reached as client.Domains.
type Domains struct {
	http *transport
}

// List returns your sending domains and their verification status.
func (d *Domains) List(ctx context.Context) ([]DomainListItem, error) {
	var out []DomainListItem
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/", nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// Create registers a new sending domain and returns the DNS records to publish.
func (d *Domains) Create(ctx context.Context, name string) (*Domain, error) {
	var out Domain
	if err := d.http.doRequest(ctx, "POST", "/v1/domains/", map[string]string{"name": name}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get fetches a domain with its DKIM selector and DNS records.
func (d *Domains) Get(ctx context.Context, id string) (*Domain, error) {
	var out Domain
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/"+url.PathEscape(id), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete deletes a domain.
func (d *Domains) Delete(ctx context.Context, id string) error {
	return d.http.doRequest(ctx, "DELETE", "/v1/domains/"+url.PathEscape(id), nil, nil)
}

// Verify re-checks DNS and verifies the domain.
func (d *Domains) Verify(ctx context.Context, id string) (*Domain, error) {
	var out Domain
	if err := d.http.doRequest(ctx, "POST", "/v1/domains/"+url.PathEscape(id)+"/verify", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Health runs live DNS health checks (DKIM, SPF, DMARC, return-path, MX).
func (d *Domains) Health(ctx context.Context, id string) (*DomainHealth, error) {
	var out DomainHealth
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/"+url.PathEscape(id)+"/health", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Diagnose diagnoses configuration issues and returns a health score.
func (d *Domains) Diagnose(ctx context.Context, id string) (*DomainDiagnosis, error) {
	var out DomainDiagnosis
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/"+url.PathEscape(id)+"/diagnose", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// MxStatus returns the current MX status (shape varies by provider).
func (d *Domains) MxStatus(ctx context.Context, id string) (map[string]any, error) {
	var out map[string]any
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/"+url.PathEscape(id)+"/mx-status", nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// PublishedRecords returns the values currently published in DNS for each of
// the domain's records (an open map).
func (d *Domains) PublishedRecords(ctx context.Context, id string) (map[string]any, error) {
	var out map[string]any
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/"+url.PathEscape(id)+"/published-records", nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// RotateDkim rotates the domain's DKIM key, returning the new record to publish.
func (d *Domains) RotateDkim(ctx context.Context, id string) (*DkimRotation, error) {
	var out DkimRotation
	if err := d.http.doRequest(ctx, "POST", "/v1/domains/"+url.PathEscape(id)+"/rotate-dkim", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Transfer initiates a transfer of this domain to another Axene account.
func (d *Domains) Transfer(ctx context.Context, id string, params TransferParams) (*DomainTransfer, error) {
	var out DomainTransfer
	if err := d.http.doRequest(ctx, "POST", "/v1/domains/"+url.PathEscape(id)+"/transfer", params, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// CheckAvailability checks whether a domain name is available to add.
func (d *Domains) CheckAvailability(ctx context.Context, name string) (*DomainAvailability, error) {
	q := url.Values{}
	q.Set("name", name)
	var out DomainAvailability
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/check-availability?"+q.Encode(), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Check checks whether a domain name already exists in your account.
func (d *Domains) Check(ctx context.Context, name string) (*DomainCheck, error) {
	var out DomainCheck
	if err := d.http.doRequest(ctx, "GET", "/v1/domains/check/"+url.PathEscape(name), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: packages/python/src/axene_mailer/resources/domains.py
"""The ``domains`` resource: register, verify, inspect, and transfer sending
domains."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from .._http import HttpTransport
from .._serialize import query


class Domains:
    """Accessed as ``axene.domains``."""

    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def list(self) -> List[Dict[str, Any]]:
        """List your sending domains and their verification status."""
        return self._http.request("GET", "/v1/domains/")

    def create(self, name: str) -> Dict[str, Any]:
        """Register a new sending domain. Returns the DNS records to publish."""
        return self._http.request("POST", "/v1/domains/", {"name": name})

    def get(self, domain_id: str) -> Dict[str, Any]:
        """Fetch a domain with its DKIM selector and DNS records."""
        return self._http.request("GET", f"/v1/domains/{domain_id}")

    def delete(self, domain_id: str) -> None:
        """Delete a domain."""
        return self._http.request("DELETE", f"/v1/domains/{domain_id}")

    def verify(self, domain_id: str) -> Dict[str, Any]:
        """Re-check DNS and verify the domain."""
        return self._http.request("POST", f"/v1/domains/{domain_id}/verify")

    def health(self, domain_id: str) -> Dict[str, Any]:
        """Run live DNS health checks (DKIM, SPF, DMARC, return-path, MX)."""
        return self._http.request("GET", f"/v1/domains/{domain_id}/health")

    def diagnose(self, domain_id: str) -> Dict[str, Any]:
        """Diagnose configuration issues and get a health score."""
        return self._http.request("GET", f"/v1/domains/{domain_id}/diagnose")

    def mx_status(self, domain_id: str) -> Dict[str, Any]:
        """Current MX status for inbound/forwarding (shape varies by provider)."""
        return self._http.request("GET", f"/v1/domains/{domain_id}/mx-status")

    def published_records(self, domain_id: str) -> Dict[str, Any]:
        """The values currently published in DNS for each of the domain's records."""
        return self._http.request("GET", f"/v1/domains/{domain_id}/published-records")

    def rotate_dkim(self, domain_id: str) -> Dict[str, Any]:
        """Rotate the domain's DKIM key, returning the new record to publish."""
        return self._http.request("POST", f"/v1/domains/{domain_id}/rotate-dkim")

    def transfer(
        self,
        domain_id: str,
        target_email: str,
        note: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Initiate a transfer of this domain to another Axene account."""
        return self._http.request(
            "POST",
            f"/v1/domains/{domain_id}/transfer",
            {"target_email": target_email, "note": note},
        )

    def check_availability(self, name: str) -> Dict[str, Any]:
        """Check whether a domain name is available to add (checks public DNS)."""
        return self._http.request("GET", f"/v1/domains/check-availability{query({'name': name})}")

    def check(self, name: str) -> Dict[str, Any]:
        """Check whether a domain name already exists in your account."""
        return self._http.request("GET", f"/v1/domains/check/{name}")
